### PR TITLE
WASI & darwin: support basic file io based on libc

### LINKED ISSUE #1505: WebAssembly binary built targeting the WASI interface cannot read from standard input
Reading from standard input from a Go binary compiled for the WASI interface is not working.
It seems the WASI backend of TinyGo doesn't have `fd_read` implemented.

# How to reproduce the issue

Create a simple program named `stdin-test.go`:

```go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	fmt.Println("say something")
	msg, err := reader.ReadString('\n')
	if err != nil {
		fmt.Printf("I got an error: %v\n", err)
	} else {
		fmt.Printf("This is what I got: |%s|\n", msg)
	}
}
```

Build the program:

```bash
$ tinygo build -o stdin-test.wasm -wasm-abi=generic -target=wasi stdin-test.go
```

Execute the program:

```bash
$ wasmtime run stdin-test.wasm 
say something
I got an error: read /dev/stdin: operation not supported
```

# Environment details

* tinygo version 0.16.0 linux/amd64 (using go version go1.15.2 and LLVM version 10.0.1)
* wasmtime 0.21.0

### PR DESCRIPTION
**Pull request by mathetake**
This is the first PR towards file system support in libc target.

resolves #1505

### CODE REVIEW

**Comment by deadprogram** [comment]
Looks like tests failing @mathetake :crying_cat_face:

File: src/os/errors.go [hunk 1]
```diff
@@ -17,8 +18,46 @@ var (
...
+	// Note that this function is not errors.Is:
+	// underlyingError only unwraps the specific error-wrapping types
+	// that it historically did, not all errors implementing Unwrap().
+	err = underlyingError(err)
+	if err == target {
+		return true
+	}
+	// To preserve prior behavior, only examine syscall errors.
+	e, ok := err.(syscall.Errno)
+	return ok && e.Is(target)
```
Comment (mathetake) [inline src/os/errors.go:63]: Looks like `e.Is` causes build failure in go 1.12 and 1.11. If we want to resolve this, the implementation may be complicated. So I'm just curious how long TinyGo will support (at least on CI testing) 1.12 and 1.11? AFAIK, official Go's support on these versions are no longer available according to the official doc saying `Each major Go release is supported until there are two newer major releases.`  https://golang.org/doc/devel/release.html#policy .

I need maintainers input. Thanks in advance.


error output: 
```
        --- FAIL: TestCompiler/ARMLinux/stdlib.go (37.42s)
            main_test.go:182: # os
            main_test.go:182: /home/circleci/project/src/os/errors.go:49:17: invalid operation: e (variable of type syscall.Errno) has no field or method Is
```

**Review by mathetake — COMMENTED** [review]
I think this is ready for review except the building failure in 1.12 and 1.11.

File: src/os/errors.go (hunk 1 above)
Comment (deadprogram) [reply src/os/errors.go:63]: That is a very valid point to consider. I just created https://github.com/tinygo-org/tinygo/issues/1672 to discuss more broadly.

Comment (mathetake) [reply src/os/errors.go:63]: Thanks, I'm going to wait until the next release is published :)

**Comment by trevor403** [comment]
@mathetake https://github.com/tinygo-org/tinygo/pull/1682 was merged, can you rebase and retest?

File: src/os/errors.go (hunk 1 above)
Comment (mathetake) [reply src/os/errors.go:63]: https://github.com/tinygo-org/tinygo/pull/1659/commits/e3bbae5fba2803027036e93ccfa6f91c014d38ae

**Comment by mathetake** [comment]
merged origin/dev and kicked CI again.

**Comment by deadprogram** [comment]
@aykevl @bradleypeabody @fgsch or anyone else who is into WebAssembly, any further comments?

@mathetake do you mind squashing your commits in this branch?

**Comment by oflebbe** [comment]
This patch does not work for me, on macos  with go 1.16
FileMode isn't defined any more .

**Comment by oflebbe** [comment]
I have no idea why https://app.circleci.com/pipelines/github/tinygo-org/tinygo/4547/workflows/9b0eae1e-04dc-429f-83dc-e7729876b030/jobs/25529  didn't fail.

Maybe it has go 1.15 installed ?

It fails for me with
  main.go:756: /Users/olaf/Devel/oss/tinygo/src/os/file.go:15:30: undeclared name: FileMode
    main.go:756: /Users/olaf/Devel/oss/tinygo/src/os/file.go:54:43: undeclared name: FileMode
    main.go:756: /Users/olaf/Devel/oss/tinygo/src/os/file.go:110:34: undeclared name: FileInfo
    main.go:756: /Users/olaf/Devel/oss/tinygo/src/os/file.go:120:24: undeclared name: FileInfo
    main.go:756: /Users/olaf/Devel/oss/tinygo/src/os/file.go:171:25: undeclared name: FileInfo
    main.go:756: /Users/olaf/Devel/oss/tinygo/src/os/file.go:176:26: undeclared name: FileInfo
    main.go:756: /Users/olaf/Devel/oss/tinygo/src/os/file_unix.go:33:50: undeclared name: FileMode
    main.go:756: /Users/olaf/Devel/oss/tinygo/src/os/file_unix.go:41:63: undeclared name: FileMode
    main.go:756: /Users/olaf/Devel/oss/tinygo/src/os/filesystem.go:31:39: undeclared name: FileMode
    main.go:756: /Users/olaf/Devel/oss/tinygo/src/os/filesystem.go:35:26: undeclared name: FileMode

**Comment by mathetake** [comment]
@oflebbe ”tinygo clean” might help you.

**Comment by oflebbe** [comment]
Indeed @mathetake  🤦🏻‍♂️. Works now.

**Comment by deadprogram** [comment]
Thanks very much for working on this @mathetake and to @oflebbe for testing.

Now merging.

**Pull request merged**